File: src/main/java/blockgame/world/generators/decorator/OreDecorator.java
package blockgame.world.generators.decorator;

import blockgame.block.BlockRegistry;
import blockgame.block.TilePalette;
import blockgame.block.TileState;
import blockgame.world.World;

import java.util.function.Predicate;

public class OreDecorator implements IDecorator {
    private final TileState _ore;
    private final int _chances;
    private final int _minY;
    private final int _maxY;
    private final int _veinSize;
    private final Predicate<TileState> _replacePredicate;
    private final boolean _scaleFail;

    public OreDecorator(TileState ore, int chances, int minY, int maxY, int veinSize, Predicate<TileState> replacePredicate,
                        boolean scaleFail) {
        this._ore = ore;
        this._chances = chances;
        this._minY = minY;
        this._maxY = maxY;
        this._veinSize = veinSize;
        this._replacePredicate = replacePredicate;
        this._scaleFail = scaleFail;
    }

    public OreDecorator(TileState ore,  int chances, int minY, int maxY, int veinSize, Predicate<TileState> replacePredicate) {
        this(ore, chances, minY, maxY, veinSize, replacePredicate, true);
    }

    public void decorate(World world, int cX, int cZ) {
        for (int c = 0;c < _chances;c++) {
            int x = world.random.nextInt(8) + 8;
            int y = world.random.nextInt(128);
            int z = world.random.nextInt(8) + 8;

            if(y < _minY || y > _maxY)
                continue;

            if(world.getTileAtAdjusted(cX, cZ, x, y, z) == null) {
                world.getTileAtAdjusted(cX, cZ, x, y, z);
            }
            if (world.getTileAtAdjusted(cX, cZ, x, y, z).block != BlockRegistry.STONE)
                continue;

            world.setTileAtAdjusted(cX, cZ, x, y, z, TilePalette.getTileRepresentation(_ore));

            int count = 1;

            while (count < _veinSize) {

                if(_scaleFail && world.random.nextInt(_veinSize) < count) {
                    break;
                }

                int dX = world.random.nextInt(3) - 1;
                int dY = world.random.nextInt(3) - 1;
                int dZ = world.random.nextInt(3) - 1;

                if (y + dY < 0 || y + dY > 128  || y + dY < _minY || y + dY > _maxY) {
                    dY = 0;
                }

                if (!_replacePredicate.test(world.getTileAtAdjusted(cX, cZ, x + dX, y +dY, z + dZ))) {
                    dX = 0;
                    dY = 0;
                    dZ = 0;
                }


                world.setTileAtAdjusted(cX, cZ, x + dX, y + dY, z + dZ, TilePalette.getTileRepresentation(_ore));
                x += dX;
                y += dY;
                z += dZ;
                count++;
            }
        }
    }
}

File: src/main/kotlin/world/generators/decorators/OreDecorator.kt
package world.generators.decorators

import block.BlockRegistration
import block.TilePalette
import block.TileState
import world.World
import kotlin.random.Random

class OreDecorator(val ore: TileState, val chances: Int, val minY: Int, val maxY: Int, val veinSize: Int, val replacePredicate: (TileState) -> Boolean,
                   val scaleFail: Boolean = true) : IDecorator {
    override fun decorate(world: World, cX: Int, cZ: Int) {
        world.random = Random(world.getSeed())
        world.random = Random(world.getSeed()
                + (world.random.nextInt(Int.MAX_VALUE) * cX)
                + (world.random.nextInt(Int.MAX_VALUE) * cZ) xor world.getSeed())
        for (c in 0 until chances) {
            var x = world.random.nextInt(0, 16)
            var y = world.random.nextInt(0, 128)
            var z = +world.random.nextInt(0, 16)

            if(y < minY || y > maxY) continue

            if (world.getTileAt((cX*16)+x, y, (cZ*16)+z)!!.block != BlockRegistration.STONE) continue
            world.setTileAt((cX*16)+x, y, (cZ*16)+z, TilePalette.getTileRepresentation(ore))

            var count = 1


            while (count < veinSize) {

                if(scaleFail && world.random.nextInt(veinSize) < count) {
                    break
                }

                var dX = world.random.nextInt(-1, 2)
                var dY = world.random.nextInt(-1, 2)
                var dZ = world.random.nextInt(-1, 2)

                if (x + dX < 0 || x + dX > 15) {
                    dX = 0
                }

                if (y + dY < 0 || y + dY > 128  || y + dY < minY || y + dY > maxY) {
                    dY = 0
                }

                if (z + dZ < 0 || z + dZ > 15) {
                    dZ = 0
                }

                if (!replacePredicate(world.getTileAt((cX*16)+x + dX, y +dY, (cZ*16)+z + dZ)!!)) {
                    dX = 0
                    dY = 0
                    dZ = 0
                }


                world.setTileAt((cX*16)+x + dX, y + dY, (cZ*16)+z + dZ, TilePalette.getTileRepresentation(ore))
                x += dX
                y += dY
                z += dZ
                count++
            }
        }
    }
}